lib/puppetserver/ca/generate_action.rb

```diff
@@ -0,0 +1,250 @@
+
+        root_key = create_private_key(private_key_length)
+        root_cert = self_signed_ca(root_key, root_ca_name, ca_ttl)
+        root_crl = create_crl_for(root_cert, root_key, ca_ttl)
+
+        int_key = create_private_key(private_key_length)
+        int_csr = create_csr(int_key, int_ca_name)
+        int_cert = sign(root_key, root_cert, int_csr, ca_ttl)
+        int_crl = create_crl_for(int_cert, int_key, ca_ttl)
+
+        Puppetserver::Utils::FileUtilities.ensure_dir(puppet.settings[:cadir])
+
+        future_files = [ { location: puppet.settings[:cacert], content: [int_cert, root_cert]},
+                         { location: puppet.settings[:cakey], content: root_key },
+                         { location: puppet.settings[:intkey], content: int_key },
+                         { location: puppet.settings[:cacrl], content: [int_crl, root_crl]}
+                       ]
+
+        future_files.each do | file|
+          @logger.warn "#{file[:location]} exists, overwriting" if File.exist?(file[:location])
+          Puppetserver::Utils::FileUtilities.write_file(file[:location], file[:content], 0640)
+        end
+      end
+
+      def create_private_key(length)
+        OpenSSL::PKey::RSA.new(length.to_i)
+      end
+
+      def self_signed_ca(key, name, ca_ttl)
+        cert = OpenSSL::X509::Certificate.new
+
+        cert.public_key = key.public_key
+        cert.subject = OpenSSL::X509::Name.new([["CN", name]])
+        cert.issuer = cert.subject
+        cert.version = 2
+        cert.serial = rand(2**128)
+
+        # Make the certificate valid as of yesterday, because so many people's
+        # clocks are out of sync.  This gives one more day of validity than people
+        # might expect, but is better than making every person who has a messed up
+        # clock fail, and better than having every cert we generate expire a day
+        # before the user expected it to when they asked for "one year".
+        cert.not_before = Time.now - (60*60*24)
+        cert.not_after  = Time.now + ca_ttl
+
+        ef = extension_factory_for(cert, cert)
+        CA_EXTENSIONS.each do |ext|
+          extension = ef.create_extension(*ext)
+          cert.add_extension(extension)
+        end
+
+        cert.sign(key, DEFAULT_SIGNING_DIGEST)
+
+        cert
+      end
+
+      def extension_factory_for(ca, cert = nil)
+        ef = OpenSSL::X509::ExtensionFactory.new
+        ef.issuer_certificate  = ca
+        ef.subject_certificate = cert if cert
+
+        ef
+      end
+
+      def create_crl_for(ca_cert, ca_key, ca_ttl)
+        crl = OpenSSL::X509::CRL.new
+        crl.version = 1
+        crl.issuer = ca_cert.subject
+
+        ef = extension_factory_for(ca_cert)
+        crl.add_extension(
+          ef.create_extension(["authorityKeyIdentifier", "keyid:always", false]))
+        crl.add_extension(
+          OpenSSL::X509::Extension.new("crlNumber", OpenSSL::ASN1::Integer(0)))
+
+        crl.last_update = Time.now
```

Comment:
Is there any value in doing something like was done above  "Time.now - (60*60*24)" ? or is Time.now sufficient?

Reply:
I know you made some updates around making this more lenient somewhere recently per Josh's feedback, was that on the agent side or the server side? If it was the agent, then `Time.now` is probably fine here.

Reply:
Puppet sets it to be Time.now - 1 second.

Reply:
ok.  I see that ( @content.last_update = time - 1)  I'll do the same.